### PR TITLE
perf(math): avoid copying twice on bigint operations

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR avoids copying twice on `math::BigInt<N>` operations

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/base/big_int.h [hunk 1]
```diff
@@ -438,47 +435,24 @@ struct ALIGNAS(internal::LimbsAlignment(N)) BigInt {
...
+    DoAdd(*this, other, carry, ret);
+    return ret;
+  }
+
   constexpr BigInt& AddInPlace(const BigInt& other) {
     uint64_t unused = 0;
     return AddInPlace(other, unused);
   }
 
   constexpr BigInt& AddInPlace(const BigInt& other, uint64_t& carry) {
```
Comment (TomTaehoonKim) [inline tachyon/math/base/big_int.h:449]:
```suggestion
  constexpr BigInt& AddInPlace(const BigInt& other, uint64_t carry) {
```
Can you make a separate commit fixing this since it's out of scope of this commit?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/base/big_int.h [hunk 2]
```diff
@@ -514,35 +512,29 @@ struct ALIGNAS(internal::LimbsAlignment(N)) BigInt {
...
+    MulResult<BigInt> result = Multiply(other);
+    BigInt<2 * N> ret;
+    memcpy(&ret[0], &result.lo[0], sizeof(uint64_t) * N);
+    memcpy(&ret[N], &result.hi[0], sizeof(uint64_t) * N);
+    return ret;
   }
 
-  constexpr BigInt& MulInPlace(const BigInt& other, BigInt& hi) {
-    BigInt lo;
+  constexpr MulResult<BigInt> Multiply(const BigInt& other) const {
```
[the remark this answers is not in the document]
Comment (Insun35) [reply tachyon/math/base/big_int.h:525]: If so, the name of the method can be changed to `Multiply` -> `DoMul` as well (for consistency w/ Add and Sub)

[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/base/big_int.h:525]: There's  a similar method called `Divide` and this is also called externally, too!

File: tachyon/math/base/big_int.h [hunk 3]
```diff
@@ -438,47 +435,24 @@ struct ALIGNAS(internal::LimbsAlignment(N)) BigInt {
     return *this;
   }
 
+  constexpr BigInt Add(const BigInt& other) {
+    uint64_t unused = 0;
+    return Add(other, unused);
+  }
+
+  constexpr BigInt Add(const BigInt& other, uint64_t& carry) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/base/big_int.h:438]: This carry is made as a reference on purpose.

**Review by Insun35 — APPROVED** [review]
LGTM

File: tachyon/math/base/big_int.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/math/base/big_int.h:525]: I see. Since `Divide()` is used in `univariate_polynomial_ops.h`, `Multiply()` could also be used externally int eh future.

File: tachyon/math/finite_fields/generator/generator_util.h [hunk 4]
```diff
@@ -30,7 +30,7 @@ std::string MpzClassToMontString(const mpz_class& v_in, const mpz_class& m_in) {
     gmp::CopyLimbs(v_in, v.limbs);
   }
 
-  BigInt<N * 2> mul_result = v.Mul(r2);
+  BigInt<N * 2> mul_result = v.MulExtend(r2);
```
Comment (ashjeong) [inline tachyon/math/finite_fields/generator/generator_util.h:33]: To standardize <2 * N>
```suggestion
  BigInt<2 * N> mul_result = v.MulExtend(r2);
```
Though there are separate instances of 2 * N and N * 2, personally, 2 * N seems better.

File: tachyon/math/finite_fields/prime_field_generic.h [hunk 5]
```diff
@@ -277,7 +277,7 @@ class PrimeField<_Config, std::enable_if_t<!_Config::kIsSpecialPrime>> final
   }
 
   constexpr PrimeField& SlowMulInPlace(const PrimeField& other) {
-    BigInt<N * 2> r = value_.Mul(other.value_);
+    BigInt<N * 2> r = value_.MulExtend(other.value_);
```
Comment (ashjeong) [inline tachyon/math/finite_fields/prime_field_generic.h:280]: To standardize <2 * N>
```suggestion
    BigInt<2 * N> r = value_.MulExtend(other.value_);

Comment (ashjeong) [reply tachyon/math/finite_fields/prime_field_generic.h:280]: Does line 280 in the file `tachyon/math/finite_fields/prime_field.h` also need to be updated?

File: tachyon/math/base/big_int.h [hunk 6]
```diff
@@ -514,35 +512,29 @@ struct ALIGNAS(internal::LimbsAlignment(N)) BigInt {
...
-  constexpr BigInt& MulInPlace(const BigInt& other, BigInt& hi) {
-    BigInt lo;
+  constexpr MulResult<BigInt> Multiply(const BigInt& other) const {
+    MulResult<BigInt> ret;
     MulResult<uint64_t> mul_result;
     FOR_FROM_SMALLEST(i, 0, N) {
       FOR_FROM_SMALLEST(j, 0, N) {
-        uint64_t& limb = (i + j) >= N ? hi.limbs[(i + j) - N] : lo.limbs[i + j];
-        mul_result = internal::u64::MulAddWithCarry(
-            limb, limbs[i], other.limbs[j], mul_result.hi);
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/base/big_int.h:581]: Yes

File: tachyon/math/finite_fields/prime_field_generic.h (hunk 5 above)
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_generic.h:280]: That was removed in https://github.com/kroma-network/tachyon/pull/363

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM
